src/lib/support/CodeUtils.h

```diff
@@ -420,9 +424,13 @@ constexpr inline const _T & max(const _T & a, const _T & b)
  *  @param[in]  aCondition  A Boolean expression to be evaluated.
  *  @param[in]  anAction    An expression or block to execute when the
  *                          assertion fails.
+ *  @param[in]  aMetricKey  An optional metric key to be logged if the condition
+ *                          evaluates to false fails. The value for the metric is
+ *                          result of the expression anAction.
  *
  */
-#define VerifyOrExit(aCondition, anAction) nlEXPECT_ACTION(aCondition, exit, anAction)
+#define VerifyOrExit(aCondition, anAction, ...)                                                                                    \
+    nlEXPECT_ACTION(aCondition, exit, LOG_METRIC_FOR_VERIFY_OR_EXIT_ACTION(anAction, ##__VA_ARGS__))
```

Comment:
So this was mostly a thought experiment from my end. What would be better for the end programmer to use these macros to accommodate tagging the metrics with least effort? There are 3 potential solutions with their own pros and cons: Here are the formats: (Arguments in [] are optional
1. VerifyOrExit(aCondition, anAction, [metricKey]): Overloaded macro that takes optional third argument. Adv is it is compatible with current format. Also better 
2. VerifyOrExit(aCondition, MATTER_LOG_METRIC(anAction, metricKey)): Requires more typing and anAction has to be evaluated when tracing is disabled. 
3. VerifyOrExitWithMetric(aCondition, anAction, metricKey): New explicit macro and requires active thought to unlearn the current macro and prefer this appropriately. 

I am ok with option 3 since it allows to expand all the remaining Return macros to add support for metrics. Will change the code to switch to option 3 since it provides consistency across the board.

Reply:
`VerifyOrExitWithMetric(kMetricKey, aCondition, anAction)` --> clearly allows tracking both the condition and metric key, even if the action is very very long (which it usually is).

Reply:
for `VerifyOrExit` the verification is a "bool" so I am unclear what the metric would log ... since we have int/uint/error_code. So say we have `VerifyOrExit(bool condition, const char *metricName)` (I ignore the actions because they seem free form) .... if I get to implement it, what do I place in the metric? If I have no value this should be an "instant" not a metric for example (just saying event happend).

Could we get a few examples of actual code and desired outcomes? it seems we want error logging on "success checks" however unsure about verify.

Reply:
Moved to explicit macros